### PR TITLE
Better time/parallelism management in manager

### PR DESCRIPTION
**Pull request by mwielgus**

### CODE REVIEW

**Comment by mwielgus** [comment]
Tests for the whole flow will come in the next PR.

File: manager/manager.go [hunk 1]
```diff
@@ -59,10 +74,12 @@ func (rm *realManager) Stop() {
 
 func (rm *realManager) Housekeep() {
 	for {
+		// Always try to get the newest metrics
+		now := time.Now()
 		start := rm.lastSync
```
Comment (piosz) [inline manager/manager.go]: Let's have a resolution size of a window. Remove last sync

File: manager/manager.go [hunk 2]
```diff
@@ -75,9 +92,27 @@ func (rm *realManager) Housekeep() {
 }
 
 func (rm *realManager) housekeep(start, end time.Time) {
-	data := rm.sources.ScrapeSources(start, end)
-	for _, p := range rm.processors {
-		data = p.Process(data)
+	if !start.Before(end) {
```
Comment (piosz) [inline manager/manager.go]: if end - start != resolution

File: manager/manager.go [hunk 3]
```diff
@@ -75,9 +92,27 @@ func (rm *realManager) Housekeep() {
...
+		glog.Warningf("Wrong time provided to housekeep start:%s end: %s", start, end)
+		return
+	}
+
+	select {
+	case <-time.After(rm.resolution / 2):
```
Comment (piosz) [inline manager/manager.go]: housekeep timeout

File: manager/manager.go [hunk 4]
```diff
@@ -75,9 +92,27 @@ func (rm *realManager) Housekeep() {
...
+		glog.Warningf("Spent too long waiting for housekeeping to start")
+		return
+
+	case <-rm.housekeepSemaphoreChan:
```
Comment (piosz) [inline manager/manager.go]: Move this up

File: manager/manager.go [hunk 5]
```diff
@@ -75,9 +92,27 @@ func (rm *realManager) Housekeep() {
...
+		// ok, good to go
 	}
-	rm.sink.ExportData(data)
+
+	go func(rm *realManager) {
+		// should always give back the semaphore
```
Comment (piosz) [inline manager/manager.go]: Move to realManager

File: manager/manager.go (hunk 1 above)
Comment (mwielgus) [inline manager/manager.go]: Done.

File: manager/manager.go (hunk 2 above)
Comment (mwielgus) [inline manager/manager.go]: This is already enforced in Housekeep. Simple sanity check is probably more than enough.

File: manager/manager.go (hunk 3 above)
Comment (mwielgus) [inline manager/manager.go]: Done.

File: manager/manager.go (hunk 4 above)
Comment (mwielgus) [inline manager/manager.go]: Done.

File: manager/manager.go (hunk 5 above)
Comment (mwielgus) [inline manager/manager.go]: We would have two functions that take start/end and do housekeeping. Lets leave it as is.

**Comment by piosz** [comment]
unit test doesn't compile. please fix

File: manager/manager.go [hunk 6]
```diff
@@ -59,14 +74,15 @@ func (rm *realManager) Stop() {
...
 func (rm *realManager) Housekeep() {
 	for {
-		start := rm.lastSync
+		// Always try to get the newest metrics
+		now := time.Now()
+		start := now.Truncate(rm.resolution)
 		end := start.Add(rm.resolution)
-		timeToNextSync := end.Sub(time.Now())
-		// TODO: consider adding some delay here
+		timeToNextSync := end.Add(scrapeOffset).Sub(end)
```
Comment (piosz) [inline manager/manager.go]: It seems to be incorrect: time to next sync will be always 5 sec

File: manager/manager.go (hunk 2 above)
Comment (piosz) [inline manager/manager.go]: OK

File: manager/manager.go (hunk 5 above)
Comment (piosz) [inline manager/manager.go]: OK

File: manager/manager.go (hunk 6 above)
Comment (mwielgus) [inline manager/manager.go]: Of course,  should be end.Add(scrapeOffset).Sub(now). Done.

**Comment by piosz** [comment]
LGTM

**Pull request merged**